cmd/src/doc.go

```diff
@@ -0,0 +1,269 @@
+		handler: handler,
+		usageFunc: func() {
+			fmt.Fprintln(flag.CommandLine.Output(), usage)
+		},
+	})
+}
+
+func openDocFile(base, fqcn string) (interface {
+	io.StringWriter
+	io.WriteCloser
+}, error) {
+	cmdPath := path.Join(append([]string{base}, strings.Split(fqcn, " ")...)...)
+	if err := os.MkdirAll(path.Dir(cmdPath), 0755); err != nil {
+		return nil, err
+	}
+
+	return os.Create(cmdPath + ".md")
+}
+
+type docRenderer struct {
+	commandTemplate *template.Template
+	groupTemplate   *template.Template
+}
+
+func newDocRenderer() (*docRenderer, error) {
+	ct, err := template.New("command").Parse(docCommandTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	gt, err := template.New("group").Parse(docGroupTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &docRenderer{
+		commandTemplate: ct,
+		groupTemplate:   gt,
+	}, nil
+}
+
+func (dr *docRenderer) RenderCommand(fqcn string, cmd *command) (string, error) {
+	flags := []*flag.Flag{}
+	cmd.flagSet.VisitAll(func(f *flag.Flag) {
+		flags = append(flags, f)
+	})
+
+	// The usage functions are, unfortunately, a bit of a mess right now. Some
+	// output to flag.CommandLine.Output(), some to os.Stderr, some to
+	// os.Stdout. So let's replace the stdout and stderr variables temporarily
+	// to capture whatever output we get, then we can put everything back after.
+	r, w, err := os.Pipe()
+	if err != nil {
+		return "", err
+	}
+	stdout := os.Stdout
+	stderr := os.Stderr
+	os.Stdout = w
+	os.Stderr = w
+	defer func() {
+		os.Stdout = stdout
+		os.Stderr = stderr
+	}()
+	out := make(chan string)
+	go func() {
+		buf := &bytes.Buffer{}
+		if _, err := io.Copy(buf, r); err != nil {
+			// This shouldn't ever panic in normal operation.
+			panic(err)
+		}
+		out <- buf.String()
+	}()
+	if cmd.usageFunc != nil {
+		cmd.usageFunc()
+	} else if cmd.flagSet.Usage != nil {
+		cmd.flagSet.Usage()
+	}
+	w.Close()
+	usage := <-out
+
+	buf := &bytes.Buffer{}
+	if err := dr.commandTemplate.Execute(buf, &docCommandContext{
+		FQCN:  fqcn,
+		Usage: usage,
+		Flags: flags,
+	}); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func (dr *docRenderer) RenderGroup(fqcn string, commands map[string]string) (string, error) {
+	buf := &bytes.Buffer{}
+	if err := dr.groupTemplate.Execute(buf, &docGroupContext{
+		FQCN:     fqcn,
+		Commands: commands,
+	}); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+type docCommandContext struct {
+	FQCN  string
+	Usage string
+	Flags []*flag.Flag
+}
+
+type docGroupContext struct {
+	FQCN     string
+	Commands map[string]string
+}
+
+const (
+	docCommandTemplate = `
+# ` + "`" + `src {{ .FQCN }}` + "`" + `
+
+{{ if .Flags }}
+## Flags
+
+| Name | Description | Default Value |
+|------|-------------|---------------|
+{{- range .Flags -}}
+{{- "\n" -}}
+| ` + "`" + `-{{ .Name }}` + "`" + ` | {{ .Usage }} | {{ if .DefValue }}` + "`" + `{{ .DefValue }}` + "`" + `{{ end }} |
+{{- end }}
+{{ end }}
+
+## Usage
+
+` + "```" + `
```

Comment:
Holy quote!

Reply:
The more I write raw strings in Go, the more I appreciate raw strings in... uh, most other languages.